Implementation of speculative devirtualization for class protocols

swiftix opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?

<!-- Description about pull request. -->

Implementation of speculative devirtualization for class protocols.

THIS IS WORK IN PROGRESS! DO NOT MERGE YET!
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

swiftix committed:
Initial version of speculative devirtualization for class protocols

swiftix commented:
@swift-ci Please test

swiftix commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -4431,7 +4431,11 @@ llvm::Value *irgen::emitHeapMetadataRefForHeapObject(IRGenFunction &IGF,
                                      suppressCast);
-
+  // Is it an opened class existential?
+  if (objectType.isAnyClassReferenceType() && isa<ArchetypeType>(objectType))
```
@rjmccall John, is this a proper way to check that emitLoadOfHeapMetadataRef could be used for class existentials? Or do I need to add something else, e.g. to check if it is not a tagged pointer or something like this?

rjmccall replied:
I actually completely forgot about tagged pointers.  Yes, you will need to check for them dynamically.

A more efficient and idiomatic way to do this check would be to dyn_cast to ArchetypeType and call requiresClass().

Note that the code you've written isn't specific to opened class existentials — it'll actually trigger on *any* class-bounded archetype.  That's probably okay, but the comment is wrong.

swiftix committed:
Add the [exact] attribute to the checked_cast_addr_br instruction.

This attribute can be used to perform exact type casts, similar to the [exact] version of the checked_cast_br.

swiftix committed:
wip

swiftix committed:
wip

swiftix committed:
Support speculative devirtualization of static methods (i.e. calls on metatypes)

swiftix committed:
Introduce the is_same_type instruction.

swiftix committed:
[sil-devirtualizer] Do not perform a speculative devirtualization for no-opt callees.

swiftix committed:
Addresses are always ABI compatible.

swiftix committed:
Add a more strict check for conversions between metatypes.

Now it is more inline with the check performed by the verifier.

swiftix committed:
Working version of speculative devirtualization of protocols.

swiftix committed:
Use checked_cast_br in EagerSpecializer.

This more efficient than using metatype instructions, because it avoids initialization of classes.

rudkx commented:
@swiftix Is this still being worked on or can we close it out?

swiftix commented:
We can close it for now. If we ever need it again, we can revive it.
